synapse/util/async_helpers.py

```diff
@@ -127,20 +136,13 @@ def observe(self) -> "defer.Deferred[_T]":
         """
         if not self._result:
             d: "defer.Deferred[_T]" = defer.Deferred()
-
-            def remove(r):
-                self._observers.discard(d)
-                return r
-
-            d.addBoth(remove)
```

Comment:
this code removed the observer from `self._observers` when it was run. I think it was redundant before, because the observer was `pop()`ed from `self._observers` anyway, but it's doubly-redundant now, since the whole of `self._observers` is thrown away.

(it was added in https://github.com/matrix-org/synapse/pull/190 - no real clues there as to why.)